pkg/proxy/proxy_test.go

```diff
@@ -0,0 +1,243 @@
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			_, err := w.Write([]byte("my response"))
+			require.NoError(s.T(), err)
+			assert.Equal(s.T(), "Bearer clusterSAToken", r.Header.Get("Authorization"))
+		}))
+		defer ts.Close()
+
+		member2, err := url.Parse(ts.URL)
+		require.NoError(s.T(), err)
+
+		fakeApp.namespaces = map[string]*namespace.Namespace{
+			"someUserID": { // noise
+				APIURL:             *member1,
+				TargetClusterToken: "",
+			},
+			userID: {
+				APIURL:             *member2,
+				TargetClusterToken: "clusterSAToken",
+			},
+		}
+
+		// when
+		client := http.Client{Timeout: 3 * time.Second}
+		resp, err := client.Do(req)
+
+		// then
+		require.NoError(s.T(), err)
+		require.NotNil(s.T(), resp)
+		assert.Equal(s.T(), http.StatusOK, resp.StatusCode)
+		s.assertResponseBody(resp, "my response")
+	})
+}
+
+func (s *TestProxySuite) TestSingleJoiningSlash() {
+	assert.Equal(s.T(), "/", singleJoiningSlash("", ""))
+	assert.Equal(s.T(), "/", singleJoiningSlash("/", "/"))
+	assert.Equal(s.T(), "/api/namespace/pods", singleJoiningSlash("", "api/namespace/pods"))
+	assert.Equal(s.T(), "proxy/", singleJoiningSlash("proxy", ""))
+	assert.Equal(s.T(), "proxy/", singleJoiningSlash("proxy", "/"))
+	assert.Equal(s.T(), "proxy/api/namespace/pods", singleJoiningSlash("proxy", "api/namespace/pods"))
+	assert.Equal(s.T(), "proxy/subpath/api/namespace/pods", singleJoiningSlash("proxy/subpath", "api/namespace/pods"))
+	assert.Equal(s.T(), "/proxy/subpath/api/namespace/pods/", singleJoiningSlash("/proxy/subpath/", "/api/namespace/pods/"))
+}
+
+func (s *TestProxySuite) request() (*http.Request, string) {
+	req, err := http.NewRequest("GET", "http://localhost:8081/api/mycoolworkspace/pods", nil)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), req)
+	userID, err := uuid.NewV4()
+	require.NoError(s.T(), err)
+	req.Header.Set("Authorization", "Bearer "+s.token(userID))
+
+	return req, userID.String()
+}
+
+func (s *TestProxySuite) token(userID uuid.UUID, extraClaims ...authsupport.ExtraClaim) string {
+	userIdentity := &authsupport.Identity{
+		ID:       userID,
+		Username: "username-" + userID.String(),
+	}
+
+	extra := append(extraClaims, authsupport.WithEmailClaim("someemail@comp.com"))
+	token, err := authsupport.GenerateSignedE2ETestToken(*userIdentity, extra...)
+	require.NoError(s.T(), err)
+
+	return token
+}
+
+func (s *TestProxySuite) assertResponseBody(resp *http.Response, expectedBody string) {
+	buf := new(bytes.Buffer)
+	_, err := buf.ReadFrom(resp.Body)
+	require.NoError(s.T(), err)
+	assert.Equal(s.T(), expectedBody, buf.String())
+}
+
+type fakeApp struct {
+	namespaces map[string]*namespace.Namespace
+	err        error
+}
+
+func (a *fakeApp) SignupService() service.SignupService {
+	panic("implement me")
+}
+
+func (a *fakeApp) VerificationService() service.VerificationService {
+	panic("implement me")
+}
```

Comment:
do we really expect that someone would implement it? I would rather say that this shouldn't be called.
```suggestion
func (a *fakeApp) SignupService() service.SignupService {
	panic("SignupService shouldn't be called")
}

func (a *fakeApp) VerificationService() service.VerificationService {
	panic("VerificationService shouldn't be called")
}
```